utils/zerotrie/src/dense.rs

```diff
@@ -0,0 +1,175 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Data structures combining [`ZeroTrie`] and [`ZeroVec`] to store data
+//! more efficiently.
+//!
+//! [`ZeroTrie`]: crate::ZeroTrie
+
+use crate::ZeroTrieSimpleAscii;
+use zerovec::ule::tuplevar::Tuple3VarULE;
+use zerovec::ule::vartuple::VarTuple;
+use zerovec::ule::vartuple::VarTupleULE;
+use zerovec::ule::EncodeAsVarULE;
+use zerovec::vecs::Index32;
+use zerovec::ZeroSlice;
+
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+#[cfg(feature = "alloc")]
+use zerovec::ZeroVec;
+
+pub(crate) type DenseType = u16;
+
+/// A data structure designed for 2-dimensional ASCII keys with a fixed
+/// delimiter that exhibit a mix of density and sparsity (owned version).
+///
+/// It stores prefixes and suffixes separated by a delimiter, which must not
+/// occur in any of the prefix or suffix strings.
+///
+/// This type can deliver data size savings if the suffixes are commonly
+/// repeated. If they are mostly unique, a simple trie is likely better.
+///
+/// # Examples
+///
+/// ```
+/// use std::collections::BTreeMap;
+/// use zerotrie::dense::ZeroAsciiDenseSparse2dTrieOwned;
+///
+/// let mut data: BTreeMap<&str, BTreeMap<&str, usize>> = BTreeMap::new();
+///
+/// data.entry("aaa").or_default().insert("BBB", 1);
+/// data.entry("aaa").or_default().insert("CCC", 2);
+/// data.entry("ddd").or_default().insert("BBB", 3);
+/// data.entry("ddd").or_default().insert("EEE", 4);
+/// data.entry("fff").or_default().insert("BBB", 5);
+///
+/// let trie = ZeroAsciiDenseSparse2dTrieOwned::try_from_btree_map_str(&data, b'/').unwrap();
+/// let trie = trie.as_borrowed();
+///
+/// assert_eq!(trie.get("aaa", "CCC"), Some(2));
+/// assert_eq!(trie.get("aaa", "EEE"), None);
+/// ```
+#[cfg(feature = "alloc")]
+#[derive(Debug, PartialEq, Eq)]
+pub struct ZeroAsciiDenseSparse2dTrieOwned {
```

Comment:
observation: this isn't generic over a Store. Making such a change would be tricky, though, you'd need to be generic over two kinds of stores. I think this is fine for now, we can experiment later.

thought: